### ISSUE TITLE
"NJS-032 Connection cannot be released" Issue in production

### ISSUE THREAD

**Issue by aralroca** [oracle/node-oracledb#566]
Just I commented in this [related issue](https://github.com/oracle/node-oracledb/issues/350) but finally I open a new issue.

I have this issue sometimes in production:

* `Fatal NI connect error 12170` (Oracle logs)
* `NJS-032 Connection cannot be released because a database call is in progress`. (Oracledb message)

And then always the node petitions return a **504 timeout** ...

I tried to put a secure Promise.all that always resolve the promise (how @JesterXL commented [here](https://github.com/oracle/node-oracledb/issues/350#issuecomment-237922805)) but the issue is still happening... 😅  

I share my Oracle service:

```javascript
import config from '../config/application';

const oracledb = require('oracledb');

oracledb.poolTimeout = 0; // Never terminate
oracledb.maxRows = 50000;

/**
 * Schema
 */
export const schema = config.oracle.schema ? `${config.oracle.schema}.` : '';

/**
 * getConnection
 */
const doConnection = () => new Promise((resolve, reject) => {
    /* eslint no-console: ["error", { allow: ["warn", "error", "log"] }] */
    oracledb.getConnection(config.oracle, (err, connection) => {
        if (err) {
            console.error('Not possible to connect to Oracle:', err.message);
            reject(err.message);
        } else {
            console.log('Connected to Oracle');
            resolve(connection);
        }
    });
});

/**
 * Close connection
 */
const closeConnection = (conn) => {
    conn.close()
        .then(() => console.log('Oracle connection closed'))
        .catch(() => console.error('Error closing the Oracle connection'));
};

/**
 * execute Query (is secure, the promise is always resolved)
 */
const executeQuery = (...args) => conn =>
    new Promise(resolve =>
        conn.execute(...args, (err, result) => resolve({ err, result, conn })));

/**
 * Query into oracledb
*/
export const dbQuery = (...args) =>
    new Promise((resolve, reject) => {
        doConnection()
            .then(executeQuery(...args))
            .then(({ err, result, conn }) => {
                if (err) {
                    reject(err.message);
                } else {
                    closeConnection(conn);
                    resolve(result.rows || result.outBinds);
                }
            })
            .catch(reject);
    });

/**
 * Query into oracledb expecting clob in the return
 */
export const dbClobQuery = (...args) =>
    new Promise((resolve, reject) => {
        doConnection()
            .then(executeQuery(...args))
            .then(({ err, result, conn }) => {
                if (err) {
                    reject(err.message);
                } else {
                    const clob = result.rows[0][0];
                    let data = '';
                    if (typeof clob.on === 'function') {
                        clob.on('data', (c) => { data += c; });
                        clob.on('end', () => {
                            closeConnection(conn);
                            resolve(JSON.parse(data));
                        });
                        clob.on('error', () => reject);
                    } else {
                        closeConnection(conn);
                        reject();
                    }
                }
            })
            .catch(reject);
    });

/**
 * Multi query with the same connection
 */
export const dbMultiQuery = (...queries) =>
    new Promise((resolve, reject) => {
        doConnection()
            .then(conn =>
                Promise.all(queries.map(q => executeQuery(...q)(conn)))
                    .then((r) => {
                        closeConnection(conn);
                        if (r.some(x => x.err)) {
                            reject(r.map(x => x.err));
                        }
                        resolve(r.map(x => x.result.rows || x.result.outBinds));
                    })
                    .catch(reject))
            .catch(reject);
    });

```

The methods that I'm using in my project are:

*  `dbQuery ` : Used to query (opening connection and close)
*  `dbClobQuery` : The same, but expecting a stream in the response (Clob)
* `dbMultiQuery`: To do multi-queries (oppening one connection and closing)

I'll appreciate if someone can check what is wrong in my code, because I'm going crazy with this issue, hehe 😅

Thanks 🙂

**Comment by cjbj** [comment]
Pro tip: make the example runnable so people are encouraged to experiment with it!

**Comment by dmcghan** [comment]
@aralroca Obviously this error is generated when you call `closeConnection `. Can you at least isolate which `closeConnection` call is responsible?

I would guess that you have a call to `dbClobQuery` that returns more than 1 row. You're code only expects one row and only processes the CLOB for that row and then tries to close the connection.

Here are some other things that stand out:
- You're not using connection pooling, but you're talking about "load". This is going to severely impact performance.
- `dbQuery` and `dbClobQuery` close connections for successful operations but not when errors are encountered. This will be more of an issue when you switch to pooling as you need to release back to the pool even on error.
- In `dbMultiQuery`, you're missing a return:
```javascript
if (r.some(x => x.err)) {
  reject(r.map(x => x.err));
  // should return here
}
resolve(r.map(x => x.result.rows || x.result.outBinds));
```

Finally, as I mentioned in the other post you pointed to, `Promise.all` will not work like people often think. The queries will be executed sequentially, not in parallel because a connection can only do one thing at a time. Using `Promise.all` is like throwing a bunch of functions to be executed over the wall (between JS and C layers). 

While using a "sureThing" can ensure you don't close a connection too early, you still end up processing all the queries. What if the error occurs in the first query? The rest shouldn't have to run since you're just going to reject anyway.

My advice would be to use async's [`eachSeries`](http://caolan.github.io/async/docs.html#eachSeries) method, which not only reads more accurately, but also provides more granular control and potentially better performance as you exit faster on error.

**Comment by dmcghan** [comment]
@aralroca You don't need to wrap `oracledb.createPool` with your own `createPool` promise wrapper. [All async functions in the driver return a promise](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#promiseoverview) if you don't pass a callback function.

I would recommend you create the pool in your main module. That way you don't have to include create pool logic in every module that wants to use it - you can just assume it's already been done. Your main module would look something like this:

```javascript
import config from '../config/application';
const oracledb = require('oracledb');

oracledb.createPool(config.oracle)
  .then(function() {
    // start the web server now...
  })
  .error(function(err) {
    // handle error
  });
```

Also, the driver now includes a [built-in pool cache](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#connpoolcache) so you don't have to assign it to a local variable. If you created the pool as above, the pool is the default pool in the cache.

Going back to your existing code (and assuming you've moved pool creation to the main module), I think you could remove `createPool`. You could also remove `doConnection` and just call `oracledb.getConnection` as it can grab connections from the default pool. 

The closeConnection function is standing out too now. It's an async function but you don't wait for it to finish before resolving or rejecting. I would use conn.close().then(resolve or reject).

Regarding the other issue, I'm not sure we've done anything to address it yet. Remember, if you're dbClobQuery gets more than 1 row back you will have a problem. 

The next thing you need to do is add some logging/debug code to your app to help identify where things are going wrong. Here's a good post on using Winston with Morgan to get the best of both worlds: http://tostring.it/2014/06/23/advanced-logging-with-nodejs/ Once that's in place you'd use the logger module instead of console.log.

**Comment by cjbj** [comment]
You can fetch multiple rows with LOBs and close connections. It's only when you are in the middle of streaming one that you can't close the connection. Generally this would give an `NJS-030: connection cannot be released because Lob operations are in progress` error.

I'm interested in knowing more about the `Fatal NI connect error 12170` and how that occurs.

Make sure you are using the latest node-oracledb.  In fact, with the LOB changes made/coming in 1.12.x, I'd test the dev release from GitHub: `npm install oracle/node-oracledb.git#v1.12.0-dev`

How big are your CLOBs?  The future node-oracledb 1.12.1-dev will have [`fetchAsString`](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#propdbfetchasstring) support for CLOBs.  This will be useful for 'small' CLOBs.  Node.js itself doesn't cope well with multi-megabyte Strings (I've seen all sorts of NAN assertion errors, out of memory errors etc at about the 256MB mark), so for large CLOBs you would want to stream them using a node-oracledb [Lob](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#lobclass) instance and handle process & discard each chunk separately.  I've found that bumping up [`lob.pieceSize`](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#proplobpiecesize) can improve streaming performance.

**Comment by cjbj** [comment]
@aralroca where are you up to with this?  Have you tried node-oracledb 1.12.2's fetchAsString support for CLOBs?
